packages/apollo-engine-reporting/src/contextualizedStats.ts

```diff
@@ -0,0 +1,246 @@
+import { DurationHistogram } from './durationHistogram';
+import {
+  IFieldStat,
+  IPathErrorStats,
+  IQueryLatencyStats,
+  IStatsContext,
+  Trace,
+  ITypeStat,
+} from 'apollo-engine-reporting-protobuf';
+
+export class QueryLatencyStats implements IQueryLatencyStats {
+  latencyCount: DurationHistogram = new DurationHistogram();
+  requestCount: number = 0;
+  cacheHits: number = 0;
+  persistedQueryHits: number = 0;
+  persistedQueryMisses: number = 0;
+  cacheLatencyCount: DurationHistogram = new DurationHistogram();
+  rootErrorStats: IPathErrorStats = Object.create(null);
+  requestsWithErrorsCount: number = 0;
+  publicCacheTtlCount: DurationHistogram = new DurationHistogram();
+  privateCacheTtlCount: DurationHistogram = new DurationHistogram();
+  registeredOperationCount: number = 0;
+  forbiddenOperationCount: number = 0;
+}
+
+export class TypeStat implements ITypeStat {
+  perFieldStat: { [k: string]: FieldStat } = Object.create(null);
+}
+
+export class FieldStat implements IFieldStat {
+  returnType: string;
+  errorsCount: number = 0;
+  count: number = 0;
+  requestsWithErrorsCount: number = 0;
+  latencyCount: DurationHistogram = new DurationHistogram();
+
+  constructor(returnType: string) {
+    this.returnType = returnType;
+  }
+}
+
+export class ContextualizedStats {
+  statsContext: IStatsContext;
+  queryLatencyStats: QueryLatencyStats;
+  perTypeStat: { [k: string]: TypeStat };
+
+  constructor(statsContext: IStatsContext) {
+    this.statsContext = statsContext;
+    this.queryLatencyStats = new QueryLatencyStats();
+    this.perTypeStat = Object.create(null);
+  }
+
+  public addTrace(trace: Trace) {
+    const queryLatencyStats = this.queryLatencyStats;
+    queryLatencyStats.requestCount++;
+    if (trace.fullQueryCacheHit) {
+      queryLatencyStats.cacheLatencyCount.incrementDuration(trace.durationNs);
+      queryLatencyStats.cacheHits++;
+    } else {
+      queryLatencyStats.latencyCount.incrementDuration(trace.durationNs);
+    }
+
+    if (
+      !trace.fullQueryCacheHit &&
+      trace.cachePolicy &&
+      trace.cachePolicy.maxAgeNs
```

Comment:
This doesn't match the QueryLatencyStats.merge logic in Kotlin-land: it checks that scope != UNKNOWN (which is done implicitly below) but doesn't rule out maxAgeNs 0. Maybe that's handled somewhere else?

BUT ALSO!!!
I ... can't find any place where trace.cachePolicy gets written. Did I accidentally never implement that???